tensorflow/python/grappler/auto_mixed_precision_test.py

```
@@ -0,0 +1,555 @@
+
+
+
+def _loop_vars_intertwined(x0, y0, functor_x, functor_y):
+  """Loop whose loop variables are intertwined."""
+  c = lambda i, j, x, y: j < 4
+  b = lambda i, j, x, y: (j+1, i+1, functor_y(y), functor_x(x))
+  init = (constant_op.constant(0), constant_op.constant(0), x0, y0)
+  ijzw = control_flow_ops.while_loop(c, b, init)
+  return ijzw
+
+
+def _lstm_cell(prev_c, prev_h, x):
+  """ LSTMCell
+  i: input gate
+  f: forget gate
+  o: output gate
+  c: cell state
+  x: input
+  h: embedding
+  """
+  bias = _bias([4])
+  w = _weight([8, 16])
+  ifoc = math_ops.matmul(array_ops.concat([x, prev_h], axis=1), w)
+  i, f, o, c = array_ops.split(ifoc, 4, axis=1)
+  i = math_ops.sigmoid(nn.bias_add(i, bias))
+  f = math_ops.sigmoid(nn.bias_add(f, bias))
+  o = math_ops.sigmoid(nn.bias_add(o, bias))
+  c = math_ops.tanh(nn.bias_add(c, bias))
+  next_c = f * prev_c + i * c
+  next_h = o * math_ops.tanh(next_c)
+  return next_c, next_h
+
+
+def _recurrent_lstm(c, h):
+  """ Dynamic single-layer LSTM with TensorArray """
+  def cond(i, c, h, ta_x):
+    del c
+    del h
+    del ta_x
+    return i < 4
+
+  def body(i, c, h, ta_x):
+    x = ta_x.read(i)
+    next_c, next_h = _lstm_cell(c, h, x)
+    return (i+1, next_c, next_h, ta_x)
+
+  ta_x = tensor_array_ops.TensorArray(
+      dtype=dtypes.float32,
+      size=4)
+  for i in range(0, 4):
+    ta_x = ta_x.write(
+        i, constant_op.constant(0.1, shape=[8, 4],
+                                dtype=dtypes.float32))
+  init = (constant_op.constant(0), c, h, ta_x)
+  r = control_flow_ops.while_loop(cond, body, init)
+  return r
+
+
+def _make_node_with_color(color, input_tensor, name=None):
+  """ Returns a node representative of the specified list type """
+  color = color.lower()
+  if color == 'w': # White node
+    weights = _weight(input_tensor.get_shape().as_list())
+    return math_ops.matmul(input_tensor, weights, name=name)
+  if color == 'g': # Gray node
+    return math_ops.sqrt(input_tensor, name=name)
```

Having sqrt and log is fine, but will cause NaNs on negative numbers, and in tests, you assert variables are all close. NaN is close to NaN but that's not a useful check, so change these to values that do not cause NaNs, or add a TODO.

---

You're right, these tests were producing NaNs and silently failing.
I've changed them to NaN-proof ops.